library/core/src/cell.rs

```diff
@@ -1811,6 +1811,37 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
 /// [`.get_mut()`]: `UnsafeCell::get_mut`
 ///
+/// `UnsafeCell<T>` has the same in-memory representation as its inner type `T`. A consequence
+/// of this guarantee is that it is possible to convert between `T` and `UnsafeCell<T>`.
+/// However, it is only valid to obtain a `*mut T` pointer or `&mut T` reference to the
+/// contents of an `UnsafeCell<T>` through [`.get()`], [`.raw_get()`] or [`.get_mut()`], e.g.:
```

Comment:
to the contents of a *shared* `UnsafeCell`, is the key here.

For fully owned `UnsafeCell`, there's no problem.

Also, the example does not demonstrate anything about the memory layout, does it? That is quite confusing given that memory layout is the topic of the paragraph. In fact the paragraph is confusing since it seems to make two barely related points.